### PR TITLE
Btree crud

### PR DESCRIPTION
**Pull request by boki1**
# Description
Implement remove and update. Refactor searching to keep the traversed path as `TreePath` and reuse it during operations.

# Test suite
```
. tools/env.sh
eu config && eu test +runtest
```

# Environment
Using gcc (GCC) 11.1.0 on a x86_64 machine with GNU/Linux.

--------------------

#### This PR makes changes which ...

- [x] are thoroughly tested
- [x] are well documented

### CODE REVIEW

File: src/core/SizeMetrics.h [hunk 1]
```diff
@@ -1,11 +1 @@
 #pragma once
```
Comment (StoyanTinchev) [inline src/core/SizeMetrics.h:1]: Why this file is empty?

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,6 @@
 [submodule "src/third-party/btree-printer"]
 	path = src/third-party/btree-printer
 	url = https://github.com/Felerius/btree-generator
+[submodule "tools/btree-generator"]
+	path = tools/btree-generator
+	url = https://github.com/Felerius/btree-generator
```
Comment (StoyanTinchev) [inline .gitmodules:6]: Equal with the previous module, I think?

File: src/core/Util.h [hunk 3]
```diff
@@ -3,47 +3,248 @@
...
+
+	constexpr auto operator<=>(const smallstr &) const noexcept = default;
+
+	friend std::ostream &operator<<(std::ostream &os, const smallstr &str) {
+		os << static_cast<const char *>(str.m_str);
+		return os;
 	}
 
-	return {};
+	constexpr const char *str() const noexcept { return m_str; }
```
Comment (StoyanTinchev) [inline src/core/Util.h:130]: Should be marked with `[[nodiscard]]`

File: src/core/storage/Pager.h [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+	NOP_STRUCTURE(StackSpaceAllocator, m_cursor);
+
+public:
+	[[nodiscard]] constexpr Position alloc() {
+		auto tmp = m_cursor;
+		m_cursor += PAGE_SIZE;
+		return tmp;
+	}
+
+	void free(const Position) {
```
Comment (StoyanTinchev) [inline src/core/storage/Pager.h:96]: Are you sure about that method?

File: src/core/storage/Pager.h [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+		throw BadAlloc{};
+	}
+
+	/// If the page is below the cursor, it is considered as allocated since there is no "official" way of freeing
+	/// pages using this allocator.
+	/// Note: The implementation does track whether the pos is actually identifing a valid page.
+	constexpr bool has_allocated(const Position pos) const noexcept {
```
Comment (StoyanTinchev) [inline src/core/storage/Pager.h:103]: should be marked `[[nodiscard]]`

File: src/core/storage/Pager.h [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            throw BadPosition(pos);
+		const auto it = std::find_if(m_freelist.begin(), m_freelist.end(), [&](const Position curr) {
+			return curr <= pos;
+		});
+        if (*it == pos && it < m_freelist.end())
+            throw BadPosition(pos);
+        m_freelist.insert(m_freelist.begin() + std::distance(m_freelist.begin(), it), pos);
+    }
+
+    [[nodiscard]] constexpr bool has_allocated(const Position) { return true; }
```
Comment (StoyanTinchev) [inline src/core/storage/Pager.h:146]: Wow, again - what is the purpose?

File: src/core/storage/Pager.h [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+	}
+
+private:
+	const std::size_t m_limit;
+	std::unordered_map<Position, CacheEntry> m_index;
+	std::list<Position> m_tracker;
+};
+
+struct LRUCache {
+	[[nodiscard]] static CacheEvictionResult evict(PageCache<LRUCache> &cache) {
```
Comment (StoyanTinchev) [inline src/core/storage/Pager.h:217]: Never used?

File: src/core/storage/Pager.h [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+private:
+	/// Read a page off a given position
+	/// Position is expected to be 1) associated with a page and, 2) the page which is allocated.
+	/// If that is not adhered to, BadRead is thrown.
+	[[nodiscard]] Page read(Position pos) {
+		if (!at_page_boundary(pos) || !m_allocator.has_allocated(pos))
+			throw BadRead{};
+
+		Page page;
+		m_disk.seekp(pos);
```
Comment (StoyanTinchev) [inline src/core/storage/Pager.h:250]: Here and 263 line are conversions from `unsigned long` to signed type `std::streamoff` (aka `long`)

File: src/core/storage/btree/Config.h [hunk 9]
```diff
@@ -3,21 +3,38 @@
...
+/// Btree configuration
+/// Contains information about the entry types (<key, val> types), pager algorithms, sizes, etc.
+/// It is expected that the used derives from this class and defines its own configuration setup for the specific use case.
 struct Config {
 	using Key = int;
 	using Val = int;
 	using Ref = int;
 
+	using PageAllocatorPolicy = StackSpaceAllocator;
+	using PageEvictionPolicy = LRUCache;
```
Comment (StoyanTinchev) [inline src/core/storage/btree/Config.h:22]: These policies aren't used

**Review by StoyanTinchev — CHANGES_REQUESTED** [review]
Overall - nice, just fix the requested changes or for some of them explain to me their purpose.

File: src/core/SizeMetrics.h (hunk 1 above)
Comment (boki1) [reply src/core/SizeMetrics.h:1]: Because we don't need it anymore. I'll delete it :smile:

File: src/core/storage/Pager.h (hunk 5 above)
Comment (boki1) [reply src/core/storage/Pager.h:103]: Yup.

File: src/core/Util.h (hunk 3 above)
Comment (boki1) [reply src/core/Util.h:130]: Yup.

File: src/core/storage/Pager.h (hunk 4 above)
Comment (boki1) [reply src/core/storage/Pager.h:96]: Yes, it needs to be implemented, in order to conform to the generic allocator API, however it is not applicable in this particular one.

File: src/core/storage/Pager.h (hunk 5 above)
Comment (boki1) [reply src/core/storage/Pager.h:103]: :+1:

File: src/core/storage/Pager.h (hunk 6 above)
Comment (boki1) [reply src/core/storage/Pager.h:146]: Hm, that's an error. I have forgotten to implement this. :smiley:

File: src/core/storage/Pager.h (hunk 7 above)
Comment (boki1) [reply src/core/storage/Pager.h:217]: It is used.

File: src/core/storage/btree/Config.h (hunk 9 above)
Comment (boki1) [reply src/core/storage/btree/Config.h:22]: They are used in `Btree.h` which defines the `Pager<>` specific type

File: src/core/Util.h (hunk 3 above)
Comment (StoyanTinchev) [reply src/core/Util.h:130]: Has not yet been marked as`[[nodiscard]]`

File: src/core/storage/Pager.h (hunk 5 above)
Comment (StoyanTinchev) [reply src/core/storage/Pager.h:103]: Has not yet been marked as `[[nodiscard]]`

File: src/core/storage/Pager.h (hunk 6 above)
Comment (StoyanTinchev) [reply src/core/storage/Pager.h:146]: When you will implement that?

**Review by StoyanTinchev — CHANGES_REQUESTED** [review]
Please, don't resolve conversations before really to commit and fix the specific bug.

**Review by StoyanTinchev — APPROVED** [review]
Please consider to add `[[nodiscard]]`

**Pull request merged**